### PR TITLE
Remove usage of InternalsVisibleTo in the internal instrumentations.

### LINKED ISSUE #1585: Do we want to have internals visible to opentelemetry-dotnet-contrib projects?
# Question

Do we want to have internals visible to opentelemetry-dotnet-contrib projects?

## Additional Context

It could make upgrading opentelemetry-dotnet-contrib easier

See open-telemetry/opentelemetry-dotnet-contrib/issues/49
and open-telemetry/opentelemetry-dotnet-contrib/issues/49

### PR DESCRIPTION
**Pull request by ejsmith**
Fixes #1585 

## Changes

This PR removes usage of `InternalsVisibleTo` within the included instrumentations. This makes it so that 3rd party instrumentations are working with the same API and can create their own instrumentations without being blocked by internal APIs. This change required adding a lot of linked source files to each instrumentation which would be nice if there weren't so many, but it does work and is something that 3rd party instrumentations can do and know that they are using unsupported APIs.

Doing this exposed that `SuppressInstrumentationScope.IncrementIfTriggered` and  `SuppressInstrumentationScope.DecrementIfTriggered` need to be part of the public API.

* [ ] Changes in public API reviewed

### CODE REVIEW

File: src/OpenTelemetry.Instrumentation.AspNet/OpenTelemetry.Instrumentation.AspNet.csproj
Comment (CodeBlanch) [inline src/OpenTelemetry.Instrumentation.AspNet/OpenTelemetry.Instrumentation.AspNet.csproj:9]: @ejsmith Is this group of includes pretty much the same in each instrumentation csproj? If so, could we move to a props file and include that instead?

Comment (ejsmith) [reply src/OpenTelemetry.Instrumentation.AspNet/OpenTelemetry.Instrumentation.AspNet.csproj:9]: They are slightly different in each, but there is probably a common base set.

File: src/OpenTelemetry.Instrumentation.AspNet/OpenTelemetry.Instrumentation.AspNet.csproj
Comment (reyang) [inline src/OpenTelemetry.Instrumentation.AspNet/OpenTelemetry.Instrumentation.AspNet.csproj:17]: Perhaps this can be moved to a common place given it is used by multiple projects?

**Comment by reyang** [comment]
Do we want the instrumentation libraries to take dependency on the SDK, or just the API?

In the initial PR #960, I suggested that we might want instrumentation library to take the convention rather than the dependency on the SDK. Happy to revisit this topic.

File: src/OpenTelemetry.Instrumentation.AspNet/OpenTelemetry.Instrumentation.AspNet.csproj
Comment (ejsmith) [reply src/OpenTelemetry.Instrumentation.AspNet/OpenTelemetry.Instrumentation.AspNet.csproj]: I've updated to move these includes to the Common.props file.

File: src/OpenTelemetry.Instrumentation.AspNet/OpenTelemetry.Instrumentation.AspNet.csproj
Comment (ejsmith) [reply src/OpenTelemetry.Instrumentation.AspNet/OpenTelemetry.Instrumentation.AspNet.csproj]: I've updated to move these includes to the Common.props file.

**Comment by ejsmith** [comment]
Just pushed an update to move the includes to the Common.props file which cleans things up pretty nicely.

**Comment by Oberon00** [comment]
> Do we want the instrumentation libraries to take dependency on the SDK, or just the API?

The spec requires that instrumentation libraries must be able to depend only on the API. Otherwise, there is no reason for the SDK/API split to exist. https://github.com/open-telemetry/opentelemetry-specification/blob/v1.0.1/specification/library-guidelines.md#requirements

> 2. Third party libraries and frameworks that add instrumentation to their code will have a dependency only on the API of OpenTelemetry client. The developers of third party libraries and frameworks do not care (and cannot know) what specific implementation of OpenTelemetry is used in the final application.

**Comment by reyang** [comment]
> > Do we want the instrumentation libraries to take dependency on the SDK, or just the API?
> 
> The spec requires that instrumentation libraries must be able to depend only on the API. Otherwise, there is no reason for the SDK/API split to exist. https://github.com/open-telemetry/opentelemetry-specification/blob/v1.0.1/specification/library-guidelines.md#requirements

Yep, the challenge here is that some instrumentation libraries **want** (not that they must) to take dependency on the SDK as they need additional features (e.g. "SuppressInstrumention" which is not an API concept). And I think it is not something I would encourage. So we might need to consider putting the "wanted" features to the API.

**Comment by reyang** [comment]
@Oberon00 I haven't been working on OTel Python for long time, do you know if there is a similar requirement in Python client, that an instrumentation library would want to suppress the spans from underlying library?

In OTel.NET, there is high demand on the following scenario:
1. Customer uses both gRPC and HTTP clients in their app (e.g. they make calls to service A which is gRPC, and to service B which is HTTP)
2. Customer is using gRPC lib + instrumentation library for gRPC calls.
3. Customer is using HttpClient + instrumentation library for HTTP calls.
4. gRPC client calls HttpClient internally, which is also instrumented automatically by the instrumentation library (mentioned in step 3).
5. Customer only wants to see the HttpClient calls that they made, rather than the calls made by gRPC client.

**Review by cijothomas — APPROVED** [review]
Thanks.

The issue about some instrumentations depending on SDK and not API is separately addressable (or not). Not blocking this PR.

**Review by reyang — APPROVED** [review]
LGTM.

**Comment by cijothomas** [comment]
I am modifying the section about "How to write instrumentation library", to reflect recent changes of eliminating ActivitySourceAdapter.
https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/docs/trace/extending-the-sdk#instrumentation-library
Will mention about API vs SDK dependency there. 
As of today, if instrumentation *require* supress-downsteam capability, it needs SDK, and it means this instrumentation is tied to *this* SDK implementation. We'll open an issue to see if  API can support suppress feature.

**Comment by reyang** [comment]
This PR is related to #893 which has a dependency on https://github.com/open-telemetry/opentelemetry-specification/issues/530 which points to https://github.com/lmolkova/oteps/blob/terminal-context/text/0069-terminal-context.md.

**Comment by ejsmith** [comment]
> This PR is related to #893 which has a dependency on [open-telemetry/opentelemetry-specification#530](https://github.com/open-telemetry/opentelemetry-specification/issues/530) which points to https://github.com/lmolkova/oteps/blob/terminal-context/text/0069-terminal-context.md.

I guess it relates to it in just that it is making 2 additional methods on `SuppressInstrumentationScope` public which were being used by the instrumentations in this project.

**Comment by CodeBlanch** [comment]
Hey just FYI there is also this thing `IgnoresAccessChecksToAttribute` which @mconnew brought up here:
https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/38#discussion_r577944123

I looked at it bit. Not much (zero?) documentation and it seems kind of heavy, but something else to consider.

**Comment by ejsmith** [comment]
> Hey just FYI there is also this thing `IgnoresAccessChecksToAttribute` which @mconnew brought up here:
> 
> https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/38#discussion_r577944123
> 
> 
> 
> I looked at it bit. Not much (zero?) documentation and it seems kind of heavy, but something else to consider.

I can live with the tests using internals. I just want the community to be able to make their own instrumentation's the same way you all make the ones that are included in the project. This PR will make sure that we can do that.

File: src/OpenTelemetry/SuppressInstrumentationScope.cs
Comment (CodeBlanch) [inline src/OpenTelemetry/SuppressInstrumentationScope.cs:90]: @ejsmith What if instead of making these two methods public we made it a partial class and moved them to a separate file so that file could be part of the include list? I bring it up just because I don't think it is a very nice API that was ever intended to be public? I think they are only called by a couple of other very internal-y spots. /cc @reyang

Comment (CodeBlanch) [reply src/OpenTelemetry/SuppressInstrumentationScope.cs:90]: Maybe won't work because it will think it's a different class? Doh

Comment (ejsmith) [reply src/OpenTelemetry/SuppressInstrumentationScope.cs:90]: Yeah, pretty sure that wouldn't work.

Comment (CodeBlanch) [reply src/OpenTelemetry/SuppressInstrumentationScope.cs:90]: @alanwest @cijothomas @reyang 

It looks like we need to promote `IncrementIfTriggered` & `DecrementIfTriggered` to the public API because they are used by `DiagnosticSourceListener`. Does anyone remember why? I just stepped through the code and it looks like `TracerProviderSdk` also does it, so we're duplicating effort.

Would this work and allow us to keep them internal to SDK?

```csharp
        public void OnNext(KeyValuePair<string, object> value)
        {
            if (!Sdk.SuppressInstrumentation) // Do nothing if suppressing.
            {
                return;
            }

            if (!this.handler.SupportsNullActivity && Activity.Current == null)
            {
                InstrumentationEventSource.Log.NullActivity(value.Key);
                return;
            }

            try
            {
                if (value.Key.EndsWith("Start", StringComparison.Ordinal))
                {
                    this.handler.OnStartActivity(Activity.Current, value.Value);
                }
                else if (value.Key.EndsWith("Stop", StringComparison.Ordinal))
                {
                    this.handler.OnStopActivity(Activity.Current, value.Value);
                }
                else if (value.Key.EndsWith("Exception", StringComparison.Ordinal))
                {
                    this.handler.OnException(Activity.Current, value.Value);
                }
                else
                {
                    this.handler.OnCustom(value.Key, Activity.Current, value.Value);
                }
            }
            catch (Exception ex)
            {
                InstrumentationEventSource.Log.UnknownErrorProcessingEvent(this.handler?.SourceName, value.Key, ex);
            }
        }
```

Comment (ejsmith) [reply src/OpenTelemetry/SuppressInstrumentationScope.cs:90]: That doesn't work because it causes the `OnStopActivity` to get skipped.

Comment (reyang) [reply src/OpenTelemetry/SuppressInstrumentationScope.cs:90]: In `TracerProviderSdk` it is more like a _nice-to-have_ thing to use `IncrementIfTriggered` & `DecrementIfTriggered` - to prevent instrumentation mistakes (failed to maintain the ref-count).
In `DiagnosticSourceListener` it is needed since this is where the _reentrant-mutex_ like feature was intended for.

We might need to find better names if we want them to be public.

Comment (ejsmith) [reply src/OpenTelemetry/SuppressInstrumentationScope.cs:90]: This may be able to just be a local flag in the DiagnosticSourceListener. I think we just need to have a flag that says skip the next stop call if start was skipped.

Comment (ejsmith) [reply src/OpenTelemetry/SuppressInstrumentationScope.cs:90]: Maybe something like this:

```csharp
                if (value.Key.EndsWith("Start", StringComparison.Ordinal))
                {
                    if (!Sdk.SuppressInstrumentation)
                    {
                        this.handler.OnStartActivity(Activity.Current, value.Value);
                    }
                    else
                    {
                        this.skipNextStopCall = true;
                    }
                }
                else if (value.Key.EndsWith("Stop", StringComparison.Ordinal))
                {
                    if (!this.skipNextStopCall)
                    {
                        this.handler.OnStopActivity(Activity.Current, value.Value);
                    }
                    else
                    {
                        this.skipNextStopCall = false;
                    }
                }
```

Comment (reyang) [reply src/OpenTelemetry/SuppressInstrumentationScope.cs:90]: Is `this.skipNextStopCall` thread safe (e.g. if multiple threads are trying to modify/access the same thing)?

Comment (ejsmith) [reply src/OpenTelemetry/SuppressInstrumentationScope.cs:90]: @reyang I was thinking that it didn't have to be because it's an instance variable for that instance of the diagnostic source listener and as far as I know a single listener isn't multi-threaded.

Comment (reyang) [reply src/OpenTelemetry/SuppressInstrumentationScope.cs:90]: I'm very confused by "a single listener isn't multi-threaded", would you elaborate?

Comment (ejsmith) [reply src/OpenTelemetry/SuppressInstrumentationScope.cs:90]: I don't know that for sure. You guys would know it better. But I'd say that if they were multithreaded then I'm not sure how the existing counter approach would work. What happens if I get one start call that is suppressed and then one that isn't and then the stop is called for the unsuppressed one before the original one that is suppressed?

Comment (ejsmith) [reply src/OpenTelemetry/SuppressInstrumentationScope.cs:90]: It seems like it might be better to check Sdk.SuppressInstrumentation on start calls and then add the activity id to a list of ignored activities and then remove from that list when stop is called. That way you can ensure that the suppression check only happens on start and properly suppresses only the matching stop events.

Comment (ejsmith) [reply src/OpenTelemetry/SuppressInstrumentationScope.cs:90]: @reyang thoughts?

Comment (reyang) [reply src/OpenTelemetry/SuppressInstrumentationScope.cs:90]: Sdk.SuppressInstrumentation is a context-aware variable instead of a global variable. I don't think any global flag would work given the nature of the free threading model we have to deal with in the instrumentation library.

Comment (ejsmith) [reply src/OpenTelemetry/SuppressInstrumentationScope.cs:90]: @reyang do you not think this is an issue overall with the current approach of suppression with incrementing and decrementing a counter? If this is all multi-threaded then you could easily get out of order start and stop calls and end up suppressing the wrong stop call.

Comment (reyang) [reply src/OpenTelemetry/SuppressInstrumentationScope.cs:90]: I don't think the current approach has such issue.

The current approach is a context aware variable https://github.com/open-telemetry/opentelemetry-dotnet/blob/29995ec245748e70293afeeefb41ec38e898c8e4/src/OpenTelemetry/SuppressInstrumentationScope.cs#L29.

It is by default using AsyncLocal, or Remoting CallContext (on legacy runtimes that don't support AsyncLocal).

Comment (ejsmith) [reply src/OpenTelemetry/SuppressInstrumentationScope.cs:90]: Ok, yeah, that makes sense. So your counter is specific to the async call stack and therefore wouldn't have multi-threading issues with starts and stops.

Well, what is the plan here? Are we going to make those APIs public or are we coming up with something different? I'd really like to get this PR finished up.

Comment (reyang) [reply src/OpenTelemetry/SuppressInstrumentationScope.cs:90]: Personally I think it makes sense to have these functionality moved to the API package (and find better names).
This is a good topic for the SIG meeting.

Comment (cijothomas) [reply src/OpenTelemetry/SuppressInstrumentationScope.cs:90]: If we promote the Increment/Decrement to public (as done in this PR).
It will force any instrumentations which need to offer "SuppressDownsteam" functionality to take SDK dependency as opposed to API dependency. This means, these instrumentations will only work for *this* particular SDK implementation, and nothing else. If the Supress instrumentation is moved to API, that also doesn't guarantee an alternate SDK implementation will support it, as there is no spec requirement for it, leaving the instrumentations to not work in alternate SDK implementation.

This issue is likely to get more common (as more and more instrumentations gets written), and is a good candidate to pursue a spec for. Once this is spec-ed out, we can expose this in the API (even pursue a change in .NET runtime to support). Until then, leaving this in SDK seems like a good alternate.

@reyang @CodeBlanch @alanwest Please comment if you have objections/alternate proposal. Will check this in next SIG meeting.

Comment (ejsmith) [reply src/OpenTelemetry/SuppressInstrumentationScope.cs:90]: I guess an alternative is that the the `DiagnosticSourceListener` can have it's own `RuntimeContextSlot` that is incremented when an activity start is skipped and decremented when the corresponding stop is skipped. That would remove the requirement of having these API's (`IncrementIfTriggered` and `DecrementIfTriggered`) made public.

Comment (ejsmith) [reply src/OpenTelemetry/SuppressInstrumentationScope.cs:90]: @cijothomas I've updated the DiagnosticSourceListener in the contrib project to use this method as an experiment to see if it worked. What do you think about this? https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/a940525eae6c320913d3433e4bc3558c4a01f67f/src/OpenTelemetry.Contrib.Shared/DiagnosticSourceInstrumentation/DiagnosticSourceListener.cs

Comment (cijothomas) [reply src/OpenTelemetry/SuppressInstrumentationScope.cs:90]: This is also technically same as taking a dependency on *this* SDK, as you are counting on the context entry name. Another alternate SDK implementation may not use the same name, or may not even support this feature.

Comment (ejsmith) [reply src/OpenTelemetry/SuppressInstrumentationScope.cs:90]: It's technically the same, but it's not requiring you to make additional methods public that you don't seem to want to make public.

Comment (ejsmith) [reply src/OpenTelemetry/SuppressInstrumentationScope.cs:90]: @cijothomas also, think you missed it, but it's using a different named slot. Not the one that does the internal suppression.

Comment (CodeBlanch) [reply src/OpenTelemetry/SuppressInstrumentationScope.cs:90]: I think that approach is reasonable @ejsmith.

Comment (ejsmith) [reply src/OpenTelemetry/SuppressInstrumentationScope.cs:90]: If we like that approach, then I will update this PR and remove the new public APIs from it. @cijothomas thoughts?

Comment (cijothomas) [reply src/OpenTelemetry/SuppressInstrumentationScope.cs:90]: thanks @ejsmith for clarifying. I'll come back to this by end of today.
With the recent change to tracerprovider to natively process legacy activities, we may not need the DiagnosticSourceListener to do the Increment/Decrement. Checking that..

Comment (ejsmith) [reply src/OpenTelemetry/SuppressInstrumentationScope.cs:90]: Pretty sure it's still needed unless something has changed recently. The ES client in the contrib project is already making use of `AddLegacyActivity`.

**Review by cijothomas — APPROVED** [review]

File: src/OpenTelemetry/SuppressInstrumentationScope.cs
Comment (cijothomas) [reply src/OpenTelemetry/SuppressInstrumentationScope.cs]: @ejsmith I have checked the most recent code. We need the DiagnosticListener to do the inc/dec counter to achieve this. `TracerProviderSdk` alone won't cover the scenario in the instrumentations. So this(https://github.com/open-telemetry/opentelemetry-dotnet/pull/1856#discussion_r589589947) approach is the best way forward now.

**Comment by ejsmith** [comment]
@cijothomas this is broken now because the instrumentations can't call the increment and decrement methods any more. I'm trying to see if I can just remove them, but that breaks the tests. Also, if I make it use the approach that I am doing from contrib where it's using it's own `RuntimeContextSlot`, it also breaks tests.

**Comment by ejsmith** [comment]
I pushed the change using the dedicated `RuntimeContextSlot` even though it's breaking tests. Hoping someone might know why it broke all the tests.

**Comment by ejsmith** [comment]
Oh fun. So this is one of those where tests fail when all run together, but if you run a small group at a time they pass. So I'm guessing that somehow the `RuntimeContextSlot` is being cross contaminated from other tests running at the same time which is obviously an issue. Any ideas as to why that might be?

**Comment by cijothomas** [comment]
> I pushed the change using the dedicated `RuntimeContextSlot` even though it's breaking tests. Hoping someone might know why it broke all the tests.

https://github.com/open-telemetry/opentelemetry-dotnet/pull/1892 I tried the same in standalone PR.

**Comment by cijothomas** [comment]
> Oh fun. So this is one of those where tests fail when all run together, but if you run a small group at a time they pass. So I'm guessing that somehow the `RuntimeContextSlot` is being cross contaminated from other tests running at the same time which is obviously an issue. Any ideas as to why that might be?

I'll check.

**Comment by ejsmith** [comment]
So I guess the difference between your PR and mine is that I have a copy of the `DiagnosticSourceListener` in each instrumentation project and each having their own instance of `RuntimeContextSlot` pointing to the same named slot.

**Comment by ejsmith** [comment]
Changed to add a random suffix to the end of the slot name, but this is probably not a great solution as it would add a lot of AsyncLocal's

File: src/OpenTelemetry/SuppressInstrumentationScope.cs
Comment (cijothomas) [reply src/OpenTelemetry/SuppressInstrumentationScope.cs]: Sorry, I correct my comment above. 
As TracerProviderSdk now entirely owns Start/Stop of activity and feeding to processor pipeline, we were able to avoid this check from DSListener as done in this PR: https://github.com/open-telemetry/opentelemetry-dotnet/pull/1893

This should unblock this PR.

**Comment by ejsmith** [comment]
@cijothomas this has been updated to only remove the internalsvisibleto now. Hopefully we can get this merged now and ensure that no other cheats for the internal instrumentations creep into the library.

File: build/Common.props
Comment (cijothomas) [inline build/Common.props:65]: could we split this into 2 conditions:
1.  IncludeDiagnosticSourceSubscriptionHelpers - require for those instrumentations which are dealing with DS based instrumentation. HttpClient/AspNet etc. This contains code specific to subscribing to DS.
2. IncludeInstrumentationHelpers - these are general Activity related helpers and are required for all instrumentations. eg; Redis which requires this, but does not require DS helpers.

Comment (ejsmith) [reply build/Common.props:65]: Yeah, I can do that. Seems unnecessary to me though. They are just extra source files, they aren't hurting anything. Just making it so that libs will need to manage 2 options instead of one.

Comment (ejsmith) [reply build/Common.props]: Done

Comment (cijothomas) [reply build/Common.props]: Thanks! I asked for it make it very clear that only some instrumentations require the DS related stuff. There were lot of confusion earlier about the need for the DS.* classes. Now its clear thats its only required, if you are dealing with a library already producing DS events, which is only a subset of libraries.
(will be updated the now incorrect section https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/docs/trace/extending-the-sdk#writing-own-instrumentation-library here)

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/OpenTelemetry.Instrumentation.StackExchangeRedis.csproj
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.StackExchangeRedis/OpenTelemetry.Instrumentation.StackExchangeRedis.csproj:10]: do we need SDK dependency in Redis?

File: build/Common.props
Comment (ejsmith) [reply build/Common.props]: Ok, are the changes I made ok?

File: src/OpenTelemetry.Instrumentation.StackExchangeRedis/OpenTelemetry.Instrumentation.StackExchangeRedis.csproj
Comment (ejsmith) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/OpenTelemetry.Instrumentation.StackExchangeRedis.csproj:10]: I'm not sure if it's needed or not, but it wasn't changed as part of this PR.

Comment (cijothomas) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/OpenTelemetry.Instrumentation.StackExchangeRedis.csproj:10]: It did not have SDK dependency before.

Comment (ejsmith) [reply src/OpenTelemetry.Instrumentation.StackExchangeRedis/OpenTelemetry.Instrumentation.StackExchangeRedis.csproj]: Hmm. My bad. Removed.

**Review by cijothomas — APPROVED** [review]
ready to go!
Thanks for continuous effort in driving this cleanups!

**Pull request merged**
